profiler/internal/cmemprof/dynamic_wrap_linux.c

```diff
@@ -0,0 +1,215 @@
+// (specified at https://refspecs.linuxfoundation.org/elf/elf.pdf). We call ELF
+// files "objects". An ELF object contains code and other data needed for running
+// programs. This information is divided up in two ways:
+//
+//	* sections, which contain distinct kinds of information such as code,
+//	  strings and defined constants, names of libraries to link, etc.
+//	* segments, which are regions of the object that must be loaded into
+//	  memory to actually run the program or make the shared library's
+//	  code and data accessible to other programs and libraries. Segments
+//	  can contain many sections, but not all sections go in a segment.
+//
+// The segments in an ELF object are defined by its program headers. At runtime,
+// we can find the program headers for the program and each linked library by
+// calling dl_iterate_phdr.
+//
+// The segment we want is PT_DYNAMIC. This segment contains references to various
+// other parts of the ELF object needed as part of dynamic linking, including:
+//
+//	* a (dynamic) symbol table that specifies functions and other data
+//	  that the object either contains (and can provide to other objects)
+//	  or needs from another object.
+//	* a string table. Symbols in the symbol table have names, and the names
+//	  are contained in the string table.
+//	* relocation tables. There are several varieties, but essentially all
+//	  of them point to symbols in the symbol table that must be resolved
+//	  by the dynamic linker.
+//
+// To summarize what we need to do:
+//
+//	* Find the shared libraries linked by the program
+//	* Locate the dynamic symbol and string tables, and the relocation tables.
+//	* Find relocation entries that point to symbols named "malloc", "calloc",
+//	  etc.
+//	* For each of those entries, edit the address of the symbol to point to
+//	  our wrappers.
+
+extern void *__wrap_malloc(size_t size);
+extern void *__wrap_calloc(size_t nmemb, size_t size);
+extern void *__wrap_realloc(void *p, size_t size);
+extern void *__wrap_valloc(size_t size);
+extern void *__wrap_aligned_alloc(size_t alignment, size_t size);
+extern int __wrap_posix_memalign(void **p, size_t alignment, size_t size);
+
+struct alloc_func {
+	char *name;
+	void *addr;
+};
+
+struct alloc_func alloc_funcs[] = {
+	{.name="malloc", .addr=&__wrap_malloc},
+	{.name="calloc", .addr=&__wrap_calloc},
+	{.name="realloc", .addr=&__wrap_realloc},
+	{.name="valloc", .addr=&__wrap_valloc},
+	{.name="aligned_alloc", .addr=&__wrap_aligned_alloc},
+	{.name="posix_memalign", .addr=&__wrap_posix_memalign},
+};
+
+static void write_table_entry(ElfW(Addr) addr, void *value) {
+	// The memory page containing the relocation table entry might not be
+	// writable. Without changing the permissions for the page, the
+	// subsequent memcpy will segfault
+	void *page = (void *)(addr & ~(0x1000 - 1));
+	mprotect(page, 0x1000, PROT_READ | PROT_WRITE);
+	memcpy((void *)(addr), &value, sizeof(void *));
+}
```

Comment:
NIT: Make this more readable, and maybe define 4096 as a const of page size (or see if we can get the page size from somewhere without hard coding it)

```diff
diff --git a/profiler/internal/cmemprof/dynamic_wrap_linux.c b/profiler/internal/cmemprof/dynamic_wrap_linux.c
index 53991e03..5b54531e 100644
--- a/profiler/internal/cmemprof/dynamic_wrap_linux.c
+++ b/profiler/internal/cmemprof/dynamic_wrap_linux.c
@@ -73,8 +73,8 @@ static void write_table_entry(ElfW(Addr) addr, void *value) {
        // The memory page containing the relocation table entry might not be
        // writable. Without changing the permissions for the page, the
        // subsequent memcpy will segfault
-       void *page = (void *)(addr & ~(0x1000 - 1));
-       mprotect(page, 0x1000, PROT_READ | PROT_WRITE);
+       void *page = (void *)((addr / 4096) * 4096);
+       mprotect(page, 4096, PROT_READ | PROT_WRITE);
        memcpy((void *)(addr), &value, sizeof(void *));
 }
```